### PR TITLE
Elements

### PR DESCRIPTION
**Pull request by patriciobcs**
- Added a new structure "Elements", to pass the values inside expression
- Optimized the operation, now constant operations are calculated on compiling time
- Add module operation
- Add return type checking
- Solve minor bugs

### CODE REVIEW

File: compiler/ifcc.g4 [hunk 1]
```diff
@@ -2,40 +2,46 @@ grammar ifcc;
...
 whileDo: 'while' '(' expression ')' '{' content '}';
 args: (expression) (',' expression)*;
 argsDef: (TYPE VARNAME) (',' TYPE VARNAME)*;
 
-CONST: [0-9]+;
-ADDSUB: '+' | '-';
-MULTDIV: '*' | '/';
-CHAR: '\'' .? '\'';
 COMMENT: '/*' .*? '*/' -> skip;
+ONELINECOMMENT: '//' .*? '\n' -> skip;
```
Comment (M4TTRX) [inline compiler/ifcc.g4]: Idk if that's normal but you are not using this in the grammar

File: compiler/ifcc.g4 [hunk 2]
```diff
@@ -2,40 +2,46 @@ grammar ifcc;
...
-ARITH: '+' | '-' | '*' | '/' | '%';
-TYPE: 'int' | 'char';
-VARNAME: [a-zA-Z_][a-zA-Z0-9_]*; 
\ No newline at end of file
+CONST: [0-9]+;
+ADDSUB: '+' | '-';
+MULTDIVMOD: '*' | '/' | '%';
+CHAR: '\'' .? '\'';
+REF: '&';
+TYPE: 'int' | 'char' | 'float' | 'double' | 'void';
```
Comment (M4TTRX) [inline compiler/ifcc.g4:44]: why did you add float and double back?

File: compiler/CodeGenVisitor.h [hunk 3]
```diff
@@ -4,26 +4,51 @@
 #include "generated/ifccBaseVisitor.h"
 using namespace std;
 
+struct Element
```
Comment (M4TTRX) [inline compiler/CodeGenVisitor.h:15]: pls add comments to what this Element struct is and why we use it.

If you have time I think it would be nice to describe why we use Variable and Function as well

File: compiler/ifcc.g4 [hunk 4]
```diff
@@ -2,40 +2,46 @@ grammar ifcc;
...
+VARNAME: [a-zA-Z_]([a-zA-Z0-9_]| '\\u' HEX4 |'\\U' HEX4 HEX4)*;
+HEX: [0-9a-fA-F];
+HEX4: HEX HEX HEX HEX;
```
Comment (M4TTRX) [inline compiler/ifcc.g4]: Same here, I don't understand the point of these they do not seem to be used anywhere in the grammar or code

**Review by M4TTRX — APPROVED** [review]
I have some minor concerns but overall nice work!

File: compiler/ifcc.g4 (hunk 1 above)
Comment (patriciobcs) [reply compiler/ifcc.g4]: I'm using it, to skip the one-line comments

File: compiler/ifcc.g4 (hunk 2 above)
Comment (patriciobcs) [reply compiler/ifcc.g4:44]: My bad, Github Copilot sometimes write more than it needs

File: compiler/CodeGenVisitor.h (hunk 3 above)
Comment (patriciobcs) [reply compiler/CodeGenVisitor.h:15]: I'll add the descriptions

File: compiler/ifcc.g4 (hunk 4 above)
Comment (patriciobcs) [reply compiler/ifcc.g4]: That is the way used by gcc to allow special characters on the identificators (variable and functions names)
